internal/route/request_executor.go

```diff
@@ -157,64 +168,76 @@ func (r *RequestExecutor) retrieveOrCacheRequest(httpReq *http.Request, requestB
-				resp = &http.Response{
+			if httpResp == nil && jsonRPCRespBody == nil {
+				httpResp = &HttpResponse{
 					StatusCode: http.StatusOK,
-					Body:       io.NopCloser(new(bytes.Buffer)),
 				}
 				rb := *err.rb
 				rb.ID = *requestBody.ID
 				rb.JSONRPC = requestBody.JSONRPCVersion
-				respBody = &rb
+				jsonRPCRespBody = &rb
 			}
 		default:
 			return nil, nil, cached, err
 		}
 	}
 
-	if resp == nil && respBody == nil {
+	if httpResp == nil && jsonRPCRespBody == nil {
 		if val == nil {
 			return nil, nil, cached, fmt.Errorf("unexpected empty response from cache")
 		}
 
 		r.logger.Debug("cache hit", zap.Any("request", requestBody), zap.Any("value", val))
 
 		// Fill in id and jsonrpc in the respBody to match the request.
-		resp = &http.Response{
+		httpResp = &HttpResponse{
 			StatusCode: http.StatusOK,
-			Body:       io.NopCloser(new(bytes.Buffer)),
 		}
-		respBody = &jsonrpc.SingleResponseBody{
+		jsonRPCRespBody = &jsonrpc.SingleResponseBody{
 			ID:      *requestBody.ID,
 			JSONRPC: requestBody.JSONRPCVersion,
 			Result:  val,
 		}
 	}
 
-	return respBody, resp, cached, nil
+	return jsonRPCRespBody, httpResp, cached, nil
 }
 
-func (r *RequestExecutor) getResponseBody(httpReq *http.Request, requestBody jsonrpc.RequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *http.Response, error) {
+func (r *RequestExecutor) getResponseBody(httpReq *http.Request, requestBody jsonrpc.RequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *HttpResponse, error) {
 	resp, err := r.httpClient.Do(httpReq)
 
 	if err != nil {
 		r.logger.Error("Error encountered when executing request.", zap.Any("request", requestBody),
-			zap.String("upstreamID", configToRoute.ID), zap.String("response", fmt.Sprintf("%v", resp)), zap.Error(err))
-		return nil, nil, &OriginError{err}
+			zap.String("upstreamID", configToRoute.ID), zap.Error(err))
+
+		return nil, nil, &OriginError{err, "", 0}
 	}
+
+	// Body can only be read once. Read it out and put it back in the response.
+	respBodyString := util.ReadAndCopyBackResponseBody(resp)
+	httpResp := &HttpResponse{resp.StatusCode}
+
+	if resp.StatusCode >= http.StatusBadRequest {
+		r.logger.Error("Non-2xx/3xx status code encountered when executing request.", zap.Any("request", requestBody),
+			zap.String("upstreamID", configToRoute.ID), zap.String("response", respBodyString),
+			zap.Int("httpStatusCode", resp.StatusCode), zap.Error(err))
+
+		return nil, httpResp, &OriginError{nil, respBodyString, resp.StatusCode}
+	}
+
 	defer resp.Body.Close()
 
-	respBody, err := jsonrpc.DecodeResponseBody(resp)
+	jsonRPCBody, err := jsonrpc.DecodeResponseBody(resp)
 
 	if err != nil {
 		r.logger.Warn("Could not deserialize response.", zap.Any("request", requestBody),
 			zap.String("upstreamID", configToRoute.ID), zap.String("response", fmt.Sprintf("%v", resp)), zap.Error(err))
-		return nil, nil, &OriginError{err}
+		return nil, httpResp, err
```

Comment:
When we see `DecodeError` from responses, we should return the response to the customer even though we can't deserialize it.

Passing the `DecodeError` back all the way back to the front so that we can do that. 

As far as I'm aware, this isn't affecting anything right now. Just thought it was something nice to fix while I'm meddling around here.